File: src/python/problems/solve_n_queens_test.py
#https://leetcode.com/problems/n-queens

from collections import defaultdict
from core.problem_base import *

class SolveNQueens(ProblemBase):
    def Solution(self, n: int) -> list[list[str]]:
        result = []
        board = [['.'] * n for _ in range(n)]

        def placeQueens(row):
            if row >= n:
                result.append([])
                for r in range(n):
                    result[-1].append(''.join(board[r]))

            for col in range(n):
                if canPlace(row, col):
                    board[row][col] = 'Q'

                    placeQueens(row + 1)

                    board[row][col] = '.'

        def canPlace(row, col):
            for r in range(0, row):
                if board[r][col] == 'Q':
                    return False
            
            r = row
            c = col
            while r >= 0 and c >= 0:
                if board[r][c] == 'Q':
                    return False
                r -= 1
                c -= 1
            
            r = row
            c = col
            while r >= 0 and c < n:
                if board[r][c] == 'Q':
                    return False
                r -= 1
                c += 1

            return True

        placeQueens(0)

        return result
            

if __name__ == '__main__':
    TestGen(SolveNQueens) \
        .Add(lambda tc: tc.Param("n", 4).Result([[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]])) \
        .Run()

File: src/go/challenges/problems/solve_n_queens_test.go
// https://leetcode.com/problems/n-queens

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type SolveNQueens struct{}

func TestSolveNQueens(t *testing.T) {
	gen := core.TestSuite[SolveNQueens]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(4).Result([][]string{{".Q..","...Q","Q...","..Q."},{"..Q.","Q...","...Q",".Q.."}})
	}).Run(t)
}

func (SolveNQueens) Solution(n int) [][]string {
    result := [][]string{}
	board := make([][]rune, n)

	for row := 0; row < n; row++ {
		for col := 0; col < n; col++ {
			board[row] = append(board[row], '.')
		}
	}

    var placeQueen func(int)
	var canPlace func(int, int) bool

	placeQueen = func(row int) {
		if row >= n {
			result = append(result, []string{})
			for r := 0; r < n; r++ {
				result[len(result) - 1] = append(result[len(result)-1], string(board[r]))
			}
			
			return
		}

		for col := 0; col < n; col++ {
			if canPlace(row, col) {
				board[row][col] = 'Q'
				placeQueen(row + 1)
				board[row][col] = '.'
			}
		}
	}


	canPlace = func(row int, col int) bool {
		for r := row; r >= 0; r-- {
            if board[r][col] == 'Q' {
                return false
            }
        }

		r := row
		c := col
		for r >= 0 && c >= 0 {
			if board[r][c] == 'Q' {
				return false
			}
			r--
			c--
		}

		r = row
		c = col
		for r >= 0 && c < n {
			if board[r][c] == 'Q' {
				return false
			}
			r--
			c++
		}

		return true
	}

	placeQueen(0)
	
	return result
}
